drivers/flash/soc_flash_nrf.c

```diff
@@ -137,6 +144,34 @@ static int flash_nrf_read(const struct device *dev, off_t addr,
 		return 0;
 	}
 
+	/* If the currently executing image is non-secure and the address
+	 * is not in the non-secure image we will not be able to read it
+	 * directly and must instead invoke a secure service to read out
+	 * the flash.
+	 */
+#ifdef CONFIG_BUILD_WITH_TFM
+	/* We assume that image_0_nonsecure is the only non-secure
+	 * partition. This will need to be updated when it is also
+	 * possible to execute from slot 1. Partitions that are assumed to
+	 * be secure are: mcuboot, image_0, image_scratch, and storage.
+	 */
+	if (!is_within_bounds(addr, len, FLASH_AREA_OFFSET(image_0_nonsecure),
```

Comment:
Question: shouldn't we also cover the area that crosses, by len from offset, from secure into non-secure area and from non-secure area into secure?
I mean this should return error and take no action, right?
What would happen if the check here would state that a given range is not in range of secure area, because for example offset + len crosses the upper boundary, and the copy would be dumped on the `memcpy` in the line 175? Would there be some access violation or the copy would fail silently leaving bogus data in destination buffer?

Reply:
> Question: shouldn't we also cover the area that crosses, by len from offset, from secure into non-secure area and from non-secure area into secure?
> I mean this should return error and take no action, right?

If a user requested such an area tfm_platform_mem_read would return an error, and then the flash_read would return an error. Is this acceptable?

> What would happen if the check here would state that a given range is not in range of secure area, because for example offset + len crosses the upper boundary, and the copy would be dumped on the memcpy in the line 175? Would there be some access violation or the copy would fail silently leaving bogus data in destination buffer?

In such a case you would get a SecureFault IRQ. Default behaviour for secure faults is to reboot (presumably so attackers can't easily re-try attacks).

Reply:
> 
> In such a case you would get a SecureFault IRQ. Default behaviour for secure faults is to reboot (presumably so attackers can't easily re-try attacks).

I guess then that very time the copy crosses areas this should trigger SecureFault so current code is OK then.